File: python/canvastekk_workflow_sdk/exceptions.py
"""
Structured Exception Hierarchy

Provides typed exceptions that the SDK and node authors can raise.
Each exception maps to a specific HTTP status code for consistent error handling.
"""

from __future__ import annotations

from typing import Any


class NodeExecutionError(Exception):
    """Base exception for all node execution errors."""

    def __init__(
        self,
        message: str,
        *,
        error_code: str = "EXECUTION_ERROR",
        details: dict[str, Any] | None = None,
    ) -> None:
        super().__init__(message)
        self.message = message
        self.error_code = error_code
        self.details = details or {}

    def to_dict(self) -> dict[str, Any]:
        """Serialize the error to a JSON-friendly dict.

        Returns:
            Dict with ``error_code``, ``message``, and ``details`` keys.
        """
        return {
            "error_code": self.error_code,
            "message": self.message,
            "details": self.details,
        }


class NodeTimeoutError(NodeExecutionError):
    """Raised when node execution exceeds the configured timeout."""

    def __init__(
        self,
        timeout_seconds: int,
        *,
        details: dict[str, Any] | None = None,
    ) -> None:
        super().__init__(
            f"Node execution timed out after {timeout_seconds}s",
            error_code="TIMEOUT",
            details=details or {"timeout_seconds": timeout_seconds},
        )
        self.timeout_seconds = timeout_seconds


class NodeValidationError(NodeExecutionError):
    """Raised when input validation against JSON Schema fails."""

    def __init__(
        self,
        message: str,
        *,
        errors: list[dict[str, Any]] | None = None,
        details: dict[str, Any] | None = None,
    ) -> None:
        super().__init__(
            message,
            error_code="VALIDATION_ERROR",
            details=details or {},
        )
        self.errors = errors or []

    def to_dict(self) -> dict[str, Any]:
        """Serialize with an additional ``errors`` list.

        Returns:
            Dict with ``error_code``, ``message``, ``details``, and ``errors``.
        """
        result = super().to_dict()
        result["errors"] = self.errors
        return result


class NodeIOError(NodeExecutionError):
    """Raised when file I/O operations fail."""

    def __init__(
        self,
        message: str,
        *,
        path: str | None = None,
        details: dict[str, Any] | None = None,
    ) -> None:
        merged_details: dict[str, Any] = details or {}
        if path:
            merged_details["path"] = path
        super().__init__(
            message,
            error_code="IO_ERROR",
            details=merged_details,
        )
        self.path = path


class NodeConfigurationError(NodeExecutionError):
    """Raised when node configuration is invalid."""

    def __init__(
        self,
        message: str,
        *,
        details: dict[str, Any] | None = None,
    ) -> None:
        super().__init__(
            message,
            error_code="CONFIGURATION_ERROR",
            details=details or {},
        )


class WorkflowExecutionError(NodeExecutionError):
    """Raised when local workflow execution fails."""

    def __init__(
        self,
        message: str,
        *,
        node_id: str | None = None,
        details: dict[str, Any] | None = None,
    ) -> None:
        merged_details: dict[str, Any] = details or {}
        if node_id:
            merged_details["node_id"] = node_id
        super().__init__(
            message,
            error_code="WORKFLOW_EXECUTION_ERROR",
            details=merged_details,
        )
        self.node_id = node_id


class WorkflowValidationError(NodeExecutionError):
    """Raised when workflow spec validation fails."""

    def __init__(
        self,
        message: str,
        *,
        errors: list[str] | None = None,
        details: dict[str, Any] | None = None,
    ) -> None:
        super().__init__(
            message,
            error_code="WORKFLOW_VALIDATION_ERROR",
            details=details or {},
        )
        self.errors = errors or []

    def to_dict(self) -> dict[str, Any]:
        result = super().to_dict()
        result["errors"] = self.errors
        return result


class NodeOutputValidationError(NodeExecutionError):
    """Raised when output validation against JSON Schema fails."""

    def __init__(
        self,
        message: str,
        *,
        errors: list[dict[str, Any]] | None = None,
        details: dict[str, Any] | None = None,
    ) -> None:
        super().__init__(
            message,
            error_code="OUTPUT_VALIDATION_ERROR",
            details=details or {},
        )
        self.errors = errors or []

    def to_dict(self) -> dict[str, Any]:
        """Serialize with an additional ``errors`` list.

        Returns:
            Dict with ``error_code``, ``message``, ``details``, and ``errors``.
        """
        result = super().to_dict()
        result["errors"] = self.errors
        return result


ERROR_CODE_TO_HTTP_STATUS: dict[str, int] = {
    "EXECUTION_ERROR": 500,
    "TIMEOUT": 408,
    "VALIDATION_ERROR": 422,
    "OUTPUT_VALIDATION_ERROR": 422,
    "IO_ERROR": 500,
    "CONFIGURATION_ERROR": 500,
    "WORKFLOW_EXECUTION_ERROR": 500,
    "WORKFLOW_VALIDATION_ERROR": 422,
}


def get_http_status_for_error(exc: NodeExecutionError) -> int:
    """Map a :class:`NodeExecutionError` to the appropriate HTTP status code.

    Args:
        exc: A structured node execution error.

    Returns:
        Integer HTTP status code (defaults to 500 for unknown error codes).
    """
    return ERROR_CODE_TO_HTTP_STATUS.get(exc.error_code, 500)

File: typescript/src/exceptions.ts
/**
 * Base error class for all node execution errors with error code and details.
 */
export class NodeExecutionError extends Error {
  readonly errorCode: string;
  readonly details: Record<string, unknown>;

  /**
   * Creates a new NodeExecutionError.
   * @param message - Error message
   * @param options - Error options including error code and details
   */
  constructor(
    message: string,
    { errorCode = "EXECUTION_ERROR", details }: {
      errorCode?: string;
      details?: Record<string, unknown>;
    } = {},
  ) {
    super(message);
    this.name = "NodeExecutionError";
    this.errorCode = errorCode;
    this.details = details ?? {};
  }

  /**
   * Converts error to dictionary for serialization.
   * @returns Dictionary with error_code, message, and details
   */
  toDict(): Record<string, unknown> {
    return {
      error_code: this.errorCode,
      message: this.message,
      details: this.details,
    };
  }
}

/**
 * Error thrown when node execution exceeds timeout limit.
 */
export class NodeTimeoutError extends NodeExecutionError {
  readonly timeoutSeconds: number;

  /**
   * Creates a new NodeTimeoutError.
   * @param timeoutSeconds - Timeout duration in seconds
   * @param options - Additional error details
   */
  constructor(
    timeoutSeconds: number,
    { details }: { details?: Record<string, unknown> } = {},
  ) {
    super(`Node execution timed out after ${timeoutSeconds}s`, {
      errorCode: "TIMEOUT",
      details: details ?? { timeout_seconds: timeoutSeconds },
    });
    this.name = "NodeTimeoutError";
    this.timeoutSeconds = timeoutSeconds;
  }
}

/**
 * Error thrown when node input validation fails.
 */
export class NodeValidationError extends NodeExecutionError {
  readonly errors: Record<string, unknown>[];

  /**
   * Creates a new NodeValidationError.
   * @param message - Error message
   * @param options - Validation errors and additional details
   */
  constructor(
    message: string,
    { errors, details }: {
      errors?: Record<string, unknown>[];
      details?: Record<string, unknown>;
    } = {},
  ) {
    super(message, { errorCode: "VALIDATION_ERROR", details });
    this.name = "NodeValidationError";
    this.errors = errors ?? [];
  }

  /**
   * Converts error to dictionary for serialization.
   * @returns Dictionary with validation errors
   */
  toDict(): Record<string, unknown> {
    return { ...super.toDict(), errors: this.errors };
  }
}

/**
 * Error thrown when node output validation fails.
 */
export class NodeOutputValidationError extends NodeExecutionError {
  readonly errors: Record<string, unknown>[];

  /**
   * Creates a new NodeOutputValidationError.
   * @param message - Error message
   * @param options - Validation errors and additional details
   */
  constructor(
    message: string,
    { errors, details }: {
      errors?: Record<string, unknown>[];
      details?: Record<string, unknown>;
    } = {},
  ) {
    super(message, { errorCode: "OUTPUT_VALIDATION_ERROR", details });
    this.name = "NodeOutputValidationError";
    this.errors = errors ?? [];
  }

  /**
   * Converts error to dictionary for serialization.
   * @returns Dictionary with validation errors
   */
  toDict(): Record<string, unknown> {
    return { ...super.toDict(), errors: this.errors };
  }
}

/**
 * Error thrown when file I/O operations fail.
 */
export class NodeIOError extends NodeExecutionError {
  readonly path: string | null;

  /**
   * Creates a new NodeIOError.
   * @param message - Error message
   * @param options - File path and additional details
   */
  constructor(
    message: string,
    { path, details }: {
      path?: string | null;
      details?: Record<string, unknown>;
    } = {},
  ) {
    const mergedDetails: Record<string, unknown> = { ...details };
    if (path) mergedDetails.path = path;
    super(message, { errorCode: "IO_ERROR", details: mergedDetails });
    this.name = "NodeIOError";
    this.path = path ?? null;
  }
}

/**
 * Error thrown when node configuration is invalid.
 */
export class NodeConfigurationError extends NodeExecutionError {
  /**
   * Creates a new NodeConfigurationError.
   * @param message - Error message
   * @param options - Additional error details
   */
  constructor(
    message: string,
    { details }: { details?: Record<string, unknown> } = {},
  ) {
    super(message, { errorCode: "CONFIGURATION_ERROR", details });
    this.name = "NodeConfigurationError";
  }
}

/**
 * Error thrown when workflow execution fails.
 */
export class WorkflowExecutionError extends NodeExecutionError {
  readonly nodeId: string | null;

  /**
   * Creates a new WorkflowExecutionError.
   * @param message - Error message
   * @param options - Node ID and additional details
   */
  constructor(
    message: string,
    { nodeId, details }: {
      nodeId?: string | null;
      details?: Record<string, unknown>;
    } = {},
  ) {
    const mergedDetails: Record<string, unknown> = { ...details };
    if (nodeId) mergedDetails.node_id = nodeId;
    super(message, { errorCode: "WORKFLOW_EXECUTION_ERROR", details: mergedDetails });
    this.name = "WorkflowExecutionError";
    this.nodeId = nodeId ?? null;
  }
}

/**
 * Error thrown when workflow validation fails.
 */
export class WorkflowValidationError extends NodeExecutionError {
  readonly errors: string[];

  /**
   * Creates a new WorkflowValidationError.
   * @param message - Error message
   * @param options - Validation errors and additional details
   */
  constructor(
    message: string,
    { errors, details }: {
      errors?: string[];
      details?: Record<string, unknown>;
    } = {},
  ) {
    super(message, { errorCode: "WORKFLOW_VALIDATION_ERROR", details });
    this.name = "WorkflowValidationError";
    this.errors = errors ?? [];
  }

  /**
   * Converts error to dictionary for serialization.
   * @returns Dictionary with validation errors
   */
  toDict(): Record<string, unknown> {
    return { ...super.toDict(), errors: this.errors };
  }
}

/**
 * Error thrown when node registration fails.
 */
export class RegistrationError extends Error {
  readonly statusCode: number;
  readonly body: Record<string, unknown>;
  /** Engine error code from the envelope, or null when unmapped. */
  readonly code: string | null;
  /** Actionable fix suggestion derived from the envelope, or null. */
  readonly guidance: string | null;

  /**
   * Creates a new RegistrationError.
   * @param statusCode - HTTP status code
   * @param body - Response body with error details
   */
  constructor(statusCode: number, body: Record<string, unknown>) {
    const message = body.detail?.toString() ?? body.message?.toString() ?? `Registration failed with status ${statusCode}`;
    super(message);
    this.name = "RegistrationError";
    this.statusCode = statusCode;
    this.body = body;
    const derived = deriveCodeAndGuidance(body, statusCode);
    this.code = derived.code;
    this.guidance = derived.guidance;
  }
}

/**
 * Unwraps a JSON-encoded detail string (engine canonical errors).
 * Non-string or unparsable values pass through unchanged.
 */
function parseDetail(detail: unknown): unknown {
  if (typeof detail === "string") {
    try {
      return JSON.parse(detail);
    } catch {
      return detail;
    }
  }
  return detail;
}

function collectFieldMessages(body: Record<string, unknown>, statusCode: number): string[] {
  const messages: string[] = [];
  const errors = body["errors"];
  if (Array.isArray(errors)) {
    for (const item of errors) {
      if (typeof item === "object" && item !== null && item["message"]) {
        messages.push(String(item["message"]));
      } else if (typeof item === "string") {
        messages.push(item);
      }
    }
  }
  if (messages.length === 0 && statusCode === 422) {
    const detail = parseDetail(body["detail"]);
    if (Array.isArray(detail)) {
      for (const item of detail) {
        if (typeof item === "object" && item !== null) {
          const loc = Array.isArray(item["loc"])
            ? (item["loc"] as unknown[]).filter((p) => p !== "body").join(".")
            : "";
          const msg = String(item["msg"] ?? "invalid value");
          messages.push(loc ? `${loc}: ${msg}` : msg);
        } else if (typeof item === "string") {
          messages.push(item);
        }
      }
    }
  }
  return messages;
}

function deriveCodeAndGuidance(
  body: Record<string, unknown>,
  statusCode: number,
): { code: string | null; guidance: string | null } {
  const rawCode = body["error"];
  const code = typeof rawCode === "string" && rawCode.length > 0 ? rawCode : null;

  if (code === "node_version_immutable") {
    const detail = parseDetail(body["detail"]);
    const changed: string[] = [];
    if (typeof detail === "object" && detail !== null && Array.isArray((detail as Record<string, unknown>)["changed_fields"])) {
      for (const item of (detail as Record<string, unknown>)["changed_fields"] as unknown[]) {
        if (typeof item === "object" && item !== null && (item as Record<string, unknown>)["field"]) {
          changed.push(String((item as Record<string, unknown>)["field"]));
        }
      }
    }
    const fields = changed.length > 0 ? changed.join(", ") : "unknown fields";
    return {
      code,
      guidance: `This version is already published and immutable (changed: ${fields}). Bump 'version' to a higher semver and re-register.`,
    };
  }
  if (statusCode === 409) {
    return {
      code,
      guidance: "The registry rejected this version conflict. Publish a version higher than the current latest.",
    };
  }
  if (statusCode === 400 || statusCode === 422) {
    const messages = collectFieldMessages(body, statusCode);
    if (messages.length > 0) {
      return { code, guidance: `Fix the validation errors and retry: ${messages.join("; ")}` };
    }
  }
  return { code, guidance: null };
}

/**
 * Mapping of error codes to HTTP status codes.
 */
export const ERROR_CODE_TO_HTTP_STATUS: Record<string, number> = {
  EXECUTION_ERROR: 500,
  TIMEOUT: 408,
  VALIDATION_ERROR: 422,
  OUTPUT_VALIDATION_ERROR: 422,
  IO_ERROR: 500,
  CONFIGURATION_ERROR: 500,
  WORKFLOW_EXECUTION_ERROR: 500,
  WORKFLOW_VALIDATION_ERROR: 422,
};

/**
 * Converts a NodeExecutionError to its HTTP status code.
 * @param exc - Node execution error
 * @returns HTTP status code (500 default)
 */
export function getHttpStatusForError(exc: NodeExecutionError): number {
  return ERROR_CODE_TO_HTTP_STATUS[exc.errorCode] ?? 500;
}
